### PR TITLE
James 1886 - MessageIdMapper implementation for Cassandra

### PR DESCRIPTION
**Pull request by chibenwa**
Tests corrections / improvement : 
 - https://github.com/apache/james-project/commit/c9cdfbe6f1e1c84744b7e486784e213249817002
 - https://github.com/apache/james-project/commit/44a3116271f67626134f80ab0f0f5ae7765c6dba
 - https://github.com/apache/james-project/commit/af4271523323e419d69a58e8c038f5e5ee92a5b3
 - https://github.com/apache/james-project/commit/4f73e9a0f19bcd53e70f3030c8dfbcdff4977a9a

Preparation : 
 - https://github.com/apache/james-project/commit/27bd7be2b09c516aef54848efa0f0d558f4fa04f
 - https://github.com/apache/james-project/commit/52df9c877f8b814976b92ea7790e4af70a134151
 - https://github.com/apache/james-project/commit/71b8b81345d5f614e9327c3def460b4fb2996ecf
 - https://github.com/apache/james-project/commit/b0a943bb74fbefcfbd7328827f4465e84ede3eca

Small improvements of MessageMapper : 
 - https://github.com/apache/james-project/commit/8d0c5def805761f5e608c36af5952cd294c211e9
 - https://github.com/apache/james-project/commit/27d74abe99df63dcd8017243c3291160bceeba68
 - https://github.com/apache/james-project/commit/5e884bf4d131b1023a10a5c946cb13ad1e1d5aef

Then it is related to MessageIdMapper.

Choices made : 
 - Duplication of logic between MessageMapper and MessageIdManager
 - Future resolution at last moment
 - Not too much efforts on MessageMapper futures.

### CODE REVIEW

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+        return imapUidDAO.retrieve(messageId, mailboxId)
+            .thenAccept(composedMessageIds -> composedMessageIds
+                    .map(ComposedMessageIdWithMetaData::getComposedMessageId)
+                    .forEach(this::deleteIds));
+    }
+
+    @Override
+    public void delete(MessageId messageId) {
+        CassandraMessageId cassandraMessageId = (CassandraMessageId) messageId;
+        messageDAO.delete(cassandraMessageId).join();
```
Comment (aduprat) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java]: Indices should be deleted first.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+    private CompletableFuture<Void> deleteIds(ComposedMessageId composedMessageId) {
+        CassandraMessageId messageId = (CassandraMessageId) composedMessageId.getMessageId();
+        CassandraId mailboxId = (CassandraId) composedMessageId.getMailboxId();
+        return CompletableFuture.allOf(imapUidDAO.delete(messageId, mailboxId),
+            messageIdDAO.delete(mailboxId, composedMessageId.getUid()));
+    }
+
+    @Override
+
```
Comment (aduprat) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java]: extra line

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageDAO.java [hunk 3]
```diff
@@ -223,58 +169,39 @@ private UDTValue toUDT(MessageAttachment messageAttachment) {
...
     }
 
-    private Pair<MailboxMessage, Stream<MessageAttachmentById>> message(Row row, FetchType fetchType) {
+    private Pair<MailboxMessage, Stream<MessageAttachmentById>> message(Row row, List<ComposedMessageIdWithMetaData> messageIds, FetchType fetchType) {
         try {
-            ComposedMessageId messageId = retrieveComposedMessageId(messageIdFactory.of(row.getUUID(MESSAGE_ID))).join();
+            ComposedMessageIdWithMetaData messageIdWithMetaData = retrieveComposedMessageId(messageIdFactory.of(row.getUUID(MESSAGE_ID)), messageIds);
+            ComposedMessageId messageId = messageIdWithMetaData.getComposedMessageId();
 
             SimpleMailboxMessage message =
```
Comment (aduprat) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageDAO.java]: We should introduce a new type for this: it's a temporary instance as we don't yet have attachments.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxCounterDAO.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    private final PreparedStatement getMessageCountStatement;
+    private final PreparedStatement getUnseenCountStatement;
+    private final PreparedStatement incrementUnseenCountStatement;
+    private final PreparedStatement incrementMessageCountStatement;
+    private final PreparedStatement decrementUnseenCountStatement;
+    private final PreparedStatement decrementMessageCountStatement;
+
+    public CassandraMailboxCounterDAO(Session session) {
+        cassandraAsyncExecutor = new CassandraAsyncExecutor(session);
+        getMessageCountStatement = session.prepare(select(CassandraMailboxCountersTable.COUNT)
```
Comment (aduprat) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxCounterDAO.java]: extract

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxCounterDAO.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            .from(CassandraMailboxCountersTable.TABLE_NAME)
+            .where(eq(CassandraMailboxCountersTable.MAILBOX_ID, bindMarker(CassandraMailboxCountersTable.MAILBOX_ID))));
+        getUnseenCountStatement = session.prepare(select(CassandraMailboxCountersTable.UNSEEN)
```
Comment (aduprat) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxCounterDAO.java]: extract

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java [hunk 6]
```diff
@@ -122,18 +122,19 @@ public CassandraMessageIdMapper(MailboxMapper mailboxMapper, AttachmentMapper at
...
         CassandraId mailboxId = (CassandraId) mailboxMessage.getMailboxId();
-        messageDAO.save(mailboxMapper.findMailboxById(mailboxId), mailboxMessage).join();
         CassandraMessageId messageId = (CassandraMessageId) mailboxMessage.getMessageId();
         ComposedMessageIdWithMetaData composedMessageIdWithMetaData = ComposedMessageIdWithMetaData.builder()
             .composedMessageId(new ComposedMessageId(mailboxId, messageId, mailboxMessage.getUid()))
             .flags(mailboxMessage.createFlags())
             .modSeq(mailboxMessage.getModSeq())
             .build();
-        CompletableFuture.allOf(imapUidDAO.insert(composedMessageIdWithMetaData),
+        CompletableFuture.allOf(
```
Comment (aduprat) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java]: Ids should be resolved after message.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java [hunk 7]
```diff
@@ -146,22 +147,29 @@ public void save(MailboxMessage mailboxMessage) throws MailboxException {
...
+                .reduce((f1, f2) -> CompletableFuture.allOf(f1, f2))
+                .orElse(CompletableFuture.completedFuture(null)));
     }
 
     @Override
     public void delete(MessageId messageId) {
         CassandraMessageId cassandraMessageId = (CassandraMessageId) messageId;
-        messageDAO.delete(cassandraMessageId).join();
-        retrieveAndDeleteIndices(cassandraMessageId, Optional.empty()).join();
+        CompletableFuture.allOf(
```
Comment (aduprat) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java]: same comment

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 8]
```diff
@@ -103,23 +103,29 @@ public long countUnseenMessagesInMailbox(Mailbox mailbox) throws MailboxExceptio
...
-    private void deleteUsingMailboxId(ComposedMessageIdWithMetaData composedMessageIdWithMetaData) {
+    private CompletableFuture<Void> deleteUsingMailboxId(ComposedMessageIdWithMetaData composedMessageIdWithMetaData) {
         ComposedMessageId composedMessageId = composedMessageIdWithMetaData.getComposedMessageId();
         CassandraMessageId messageId = (CassandraMessageId) composedMessageId.getMessageId();
         CassandraId mailboxId = (CassandraId) composedMessageId.getMailboxId();
         MessageUid uid = composedMessageId.getUid();
-        CompletableFuture.allOf(imapUidDAO.delete(messageId, mailboxId),
-                messageIdDAO.delete(mailboxId, uid))
-            .join();
+        return CompletableFuture.allOf(imapUidDAO.delete(messageId, mailboxId),
```
Comment (aduprat) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: same

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/CassandraAsyncExecutor.java [hunk 9]
```diff
@@ -53,7 +53,7 @@ public CassandraAsyncExecutor(Session session) {
 
     public CompletableFuture<Void> executeVoid(Statement statement) {
         return FutureConverter.toCompletableFuture(session.executeAsync(statement))
-            .thenAccept(this::consume);
+            .thenApply(result -> null);
```
Comment (rouazana) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/CassandraAsyncExecutor.java]: or thenAccept(result -> {}) ?

File: mailbox/api/src/test/java/org/apache/james/mailbox/BeansTest.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+import org.apache.james.mailbox.model.ComposedMessageId;
+import org.junit.Test;
+
+import nl.jqno.equalsverifier.EqualsVerifier;
+
+public class BeansTest {
+
+    @Test
+    public void beanShouldRespectBeanContract() {
```
Comment (rouazana) [inline mailbox/api/src/test/java/org/apache/james/mailbox/BeansTest.java]: s/bean/composedMessageId/ ? And what about putting this test into ComposedMessageIdTest?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxCounterDAO.java (hunk 4 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxCounterDAO.java]: I also factorized more by using one single read statement (as we are returning integers)

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/AttachmentLoader.java [hunk 11]
```diff
@@ -40,22 +41,22 @@ public AttachmentLoader(AttachmentMapper attachmentMapper) {
         this.attachmentMapper = attachmentMapper;
     }
 
-    public Collection<org.apache.james.mailbox.model.MessageAttachment> getAttachments(Set<CassandraMessageDAO.MessageAttachment> attachmentsById) {
+    public Collection<MessageAttachment> getAttachments(Set<CassandraMessageDAO.MessageAttachmentRepresentation> attachmentsById) {
```
Comment (rouazana) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/AttachmentLoader.java]: Sorry my renaming was more about the Set<CassandraMessageDAO.MessageAttachmentRepresentation> attachmentsById stuff. Globally attachmentsById is too more used in this class, without the expected meaning.

File: mailbox/api/src/test/java/org/apache/james/mailbox/ComposedMessageIdTestBeansTest.java [hunk 12]
```diff
@@ -24,10 +24,10 @@
 
 import nl.jqno.equalsverifier.EqualsVerifier;
 
-public class BeansTest {
+public class ComposedMessageIdTestBeansTest {
```
Comment (rouazana) [inline mailbox/api/src/test/java/org/apache/james/mailbox/ComposedMessageIdTestBeansTest.java]: just ComposedMessageIdTest no?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/AttachmentLoader.java [hunk 13]
```diff
@@ -41,9 +41,9 @@ public AttachmentLoader(AttachmentMapper attachmentMapper) {
         this.attachmentMapper = attachmentMapper;
     }
 
-    public Collection<MessageAttachment> getAttachments(Set<CassandraMessageDAO.MessageAttachmentRepresentation> attachmentsById) {
+    public Collection<MessageAttachment> getAttachments(Set<CassandraMessageDAO.MessageAttachmentRepresentation> attachmentRepresentations) {
 
-        Map<AttachmentId, CassandraMessageDAO.MessageAttachmentRepresentation> attachmentIds = attachmentsById.stream()
+        Map<AttachmentId, CassandraMessageDAO.MessageAttachmentRepresentation> attachmentIds = attachmentRepresentations.stream()
```
Comment (rouazana) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/AttachmentLoader.java]: s/attachmentIds/attachmentRepresentationById/

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/AttachmentLoader.java [hunk 14]
```diff
@@ -41,9 +41,9 @@ public AttachmentLoader(AttachmentMapper attachmentMapper) {
...
                 .collect(Guavate.toImmutableMap(CassandraMessageDAO.MessageAttachmentRepresentation::getAttachmentId, Function.identity()));
 
         Map<AttachmentId, Attachment> attachmentByIdMap = attachmentsById(attachmentIds.keySet());
```
Comment (rouazana) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/AttachmentLoader.java]: s/Map// and we should be done

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java [hunk 15]
```diff
@@ -123,13 +123,14 @@ public CassandraMessageIdMapper(MailboxMapper mailboxMapper, AttachmentMapper at
     @Override
     public void save(MailboxMessage mailboxMessage) throws MailboxException {
         CassandraId mailboxId = (CassandraId) mailboxMessage.getMailboxId();
+        mailboxMapper.findMailboxById(mailboxId);
```
Comment (quynhn) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java]: For what or it just checks the mailbox is existed or not?

Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java]: It checks MBX exists

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMapperProvider.java [hunk 16]
```diff
@@ -135,12 +146,15 @@ public boolean supportPartialAttachmentFetch() {
...
     public AnnotationMapper createAnnotationMapper() throws MailboxException {
+        CassandraMessageIdToImapUidDAO imapUidDAO = new CassandraMessageIdToImapUidDAO(cassandra.getConf(), MESSAGE_ID_FACTORY);
         return new CassandraMailboxSessionMapperFactory(
                 new CassandraUidProvider(cassandra.getConf()),
                 cassandraModSeqProvider,
                 cassandra.getConf(),
                 cassandra.getTypesProvider(),
-                new DefaultMessageId.Factory()
+                new CassandraMessageDAO(cassandra.getConf(), cassandra.getTypesProvider(), MESSAGE_ID_FACTORY, imapUidDAO),
+                new CassandraMessageIdDAO(cassandra.getConf(), MESSAGE_ID_FACTORY),
```
Comment (chibenwa) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMapperProvider.java]: Idem

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/AttachmentLoader.java [hunk 17]
```diff
@@ -43,12 +43,12 @@ public AttachmentLoader(AttachmentMapper attachmentMapper) {
...
 
-        Map<AttachmentId, CassandraMessageDAO.MessageAttachmentRepresentation> attachmentIds = attachmentRepresentations.stream()
+        Map<AttachmentId, CassandraMessageDAO.MessageAttachmentRepresentation> attachmentRepresentationByIds = attachmentRepresentations.stream()
                 .collect(Guavate.toImmutableMap(CassandraMessageDAO.MessageAttachmentRepresentation::getAttachmentId, Function.identity()));
 
-        Map<AttachmentId, Attachment> attachmentByIdMap = attachmentsById(attachmentIds.keySet());
+        Map<AttachmentId, Attachment> attachmentById = attachmentsById(attachmentRepresentationByIds.keySet());
 
-        return MapMerger.merge(attachmentByIdMap, attachmentIds, this::constructMessageAttachment).values();
+        return MapMerger.merge(attachmentById, attachmentRepresentationByIds, this::constructMessageAttachment).values();
```
Comment (rouazana) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/AttachmentLoader.java]: sorry to be picky about that, but your are not consistent with the usage of final 's'
after that you can rebase and autosquash for me

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Pull request merged**